// ng-modal-lib/esm2015/lib/modal/modal-module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ModalComponent } from './modal.component';
import { ResizableModule } from '../resizable/resizable-module';
import { DraggableModule } from '../draggable/draggable-module';
import * as ɵngcc0 from '@angular/core';
export class ModalModule {
}
ModalModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ModalModule });
ModalModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ModalModule_Factory(t) { return new (t || ModalModule)(); }, imports: [[
            CommonModule,
            ResizableModule,
            DraggableModule,
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ModalModule, { declarations: function () { return [ModalComponent]; }, imports: function () { return [CommonModule,
        ResizableModule,
        DraggableModule]; }, exports: function () { return [ModalComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ModalModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    ResizableModule,
                    DraggableModule,
                ],
                declarations: [
                    ModalComponent,
                ],
                exports: [
                    ModalComponent,
                ]
            }]
    }], null, null); })();
