// ng-modal-lib/esm2015/lib/modal/modal.component.js
import { Component, ElementRef, ViewChild, Input, Output, HostListener, EventEmitter, ViewEncapsulation } from '@angular/core';
import { maxZIndex, findAncestor } from '../common/utils';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../resizable/resizable.directive';
import * as ɵngcc2 from '../draggable/draggable.directive';
import * as ɵngcc3 from '@angular/common';

const _c0 = ["modalRoot"];
const _c1 = ["modalBody"];
const _c2 = ["modalHeader"];
const _c3 = ["modalFooter"];
const _c4 = ["closeIcon"];
const _c5 = function (a0, a1) { return { "dt-icon-maximize": a0, "dt-icon-normalize": a1 }; };
function ModalComponent_i_8_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "i", 14);
    ɵngcc0.ɵɵlistener("click", function ModalComponent_i_8_Template_i_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.toggleMaximize($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(1, _c5, !ctx_r2.maximized, ctx_r2.maximized));
} }
const _c6 = [[["", 8, "app-modal-header"]], [["", 8, "app-modal-body"]], [["", 8, "app-modal-footer"]]];
const _c7 = [".app-modal-header", ".app-modal-body", ".app-modal-footer"];
export class ModalComponent {
    constructor(element) {
        this.element = element;
        this.scrollTopEnable = true;
        this.backdrop = true;
        this.closeModal = new EventEmitter();
    }
    ngAfterViewChecked() {
        if (this.executePostDisplayActions) {
            this.center();
            this.executePostDisplayActions = false;
        }
    }
    onKeyDown(event) {
        event.preventDefault();
        event.stopPropagation();
        this.hide();
    }
    onWindowResize() {
        this.executePostDisplayActions = true;
        this.center();
    }
    show() {
        this.executePostDisplayActions = true;
        this.visible = true;
        setTimeout(() => {
            this.modalRoot.nativeElement.focus();
            if (this.scrollTopEnable) {
                this.modalBody.nativeElement.scrollTop = 0;
            }
        }, 1);
    }
    hide() {
        this.visible = false;
        this.closeModal.emit(true);
        this.focusLastModal();
    }
    center() {
        let elementWidth = this.modalRoot.nativeElement.offsetWidth;
        let elementHeight = this.modalRoot.nativeElement.offsetHeight;
        if (elementWidth === 0 && elementHeight === 0) {
            this.modalRoot.nativeElement.style.visibility = 'hidden';
            this.modalRoot.nativeElement.style.display = 'block';
            elementWidth = this.modalRoot.nativeElement.offsetWidth;
            elementHeight = this.modalRoot.nativeElement.offsetHeight;
            this.modalRoot.nativeElement.style.display = 'none';
            this.modalRoot.nativeElement.style.visibility = 'visible';
        }
        const x = Math.max((window.innerWidth - elementWidth) / 2, 0);
        const y = Math.max((window.innerHeight - elementHeight) / 2, 0);
        this.modalRoot.nativeElement.style.left = x + 'px';
        this.modalRoot.nativeElement.style.top = y + 'px';
    }
    initDrag(event) {
        if (event.target === this.closeIcon.nativeElement) {
            return;
        }
        if (!this.maximized) {
            this.dragEventTarget = event;
        }
    }
    onResize(event) {
        if (event.direction === 'vertical') {
            this.calcBodyHeight();
        }
    }
    calcBodyHeight() {
        const diffHeight = this.modalHeader.nativeElement.offsetHeight + this.modalFooter.nativeElement.offsetHeight;
        const contentHeight = this.modalRoot.nativeElement.offsetHeight - diffHeight;
        this.modalBody.nativeElement.style.height = contentHeight + 'px';
        this.modalBody.nativeElement.style.maxHeight = 'none';
    }
    getMaxModalIndex() {
        return maxZIndex('.ui-modal');
    }
    focusLastModal() {
        const modal = findAncestor(this.element.nativeElement.parentElement, '.ui-modal');
        if (modal) {
            modal.focus();
        }
    }
    toggleMaximize(event) {
        if (this.maximized) {
            this.revertMaximize();
        }
        else {
            this.maximize();
        }
        event.preventDefault();
    }
    maximize() {
        this.preMaximizePageX = parseFloat(this.modalRoot.nativeElement.style.top);
        this.preMaximizePageY = parseFloat(this.modalRoot.nativeElement.style.left);
        this.preMaximizeRootWidth = this.modalRoot.nativeElement.offsetWidth;
        this.preMaximizeRootHeight = this.modalRoot.nativeElement.offsetHeight;
        this.preMaximizeBodyHeight = this.modalBody.nativeElement.offsetHeight;
        this.modalRoot.nativeElement.style.top = '0px';
        this.modalRoot.nativeElement.style.left = '0px';
        this.modalRoot.nativeElement.style.width = '100vw';
        this.modalRoot.nativeElement.style.height = '100vh';
        const diffHeight = this.modalHeader.nativeElement.offsetHeight + this.modalFooter.nativeElement.offsetHeight;
        this.modalBody.nativeElement.style.height = 'calc(100vh - ' + diffHeight + 'px)';
        this.modalBody.nativeElement.style.maxHeight = 'none';
        this.maximized = true;
    }
    revertMaximize() {
        this.modalRoot.nativeElement.style.top = this.preMaximizePageX + 'px';
        this.modalRoot.nativeElement.style.left = this.preMaximizePageY + 'px';
        this.modalRoot.nativeElement.style.width = this.preMaximizeRootWidth + 'px';
        this.modalRoot.nativeElement.style.height = this.preMaximizeRootHeight + 'px';
        this.modalBody.nativeElement.style.height = this.preMaximizeBodyHeight + 'px';
        this.maximized = false;
    }
    moveOnTop() {
        if (!this.backdrop) {
            const maxModalIndex = this.getMaxModalIndex();
            let zIndex = parseFloat(window.getComputedStyle(this.modalRoot.nativeElement).zIndex) || 0;
            if (zIndex <= maxModalIndex) {
                zIndex = maxModalIndex + 1;
                this.modalRoot.nativeElement.style.zIndex = zIndex.toString();
            }
        }
    }
}
ModalComponent.ɵfac = function ModalComponent_Factory(t) { return new (t || ModalComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ModalComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ModalComponent, selectors: [["app-modal"]], viewQuery: function ModalComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, true);
        ɵngcc0.ɵɵviewQuery(_c1, true);
        ɵngcc0.ɵɵviewQuery(_c2, true);
        ɵngcc0.ɵɵviewQuery(_c3, true);
        ɵngcc0.ɵɵviewQuery(_c4, true);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.modalRoot = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.modalBody = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.modalHeader = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.modalFooter = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.closeIcon = _t.first);
    } }, hostBindings: function ModalComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("keydown.esc", function ModalComponent_keydown_esc_HostBindingHandler($event) { return ctx.onKeyDown($event); })("resize", function ModalComponent_resize_HostBindingHandler() { return ctx.onWindowResize(); }, false, ɵngcc0.ɵɵresolveWindow);
    } }, inputs: { scrollTopEnable: "scrollTopEnable", backdrop: "backdrop", maximizable: "maximizable", inViewport: "inViewport" }, outputs: { closeModal: "closeModal" }, ngContentSelectors: _c7, decls: 17, vars: 15, consts: [[1, "ui-modal-overlay"], ["tabindex", "-1", "role", "dialog", "appResizable", "", "appDraggable", "", 1, "ui-modal", 3, "south", "east", "southEast", "southWest", "west", "northWest", "north", "northEast", "dragEventTarget", "inViewport", "resizing", "mousedown", "touchstart"], ["modalRoot", ""], [1, "ui-modal-header", 3, "mousedown", "touchstart"], ["modalHeader", ""], [1, "ui-titlebar"], [1, "ui-controlbar"], ["class", "ui-icon", 3, "ngClass", "click", 4, "ngIf"], [1, "ui-icon", "dt-icon-close", 3, "click"], ["closeIcon", ""], [1, "ui-modal-body"], ["modalBody", ""], [1, "ui-modal-footer"], ["modalFooter", ""], [1, "ui-icon", 3, "ngClass", "click"]], template: function ModalComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c6);
        ɵngcc0.ɵɵelement(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1, 2);
        ɵngcc0.ɵɵlistener("resizing", function ModalComponent_Template_div_resizing_1_listener($event) { return ctx.onResize($event); })("mousedown", function ModalComponent_Template_div_mousedown_1_listener() { return ctx.moveOnTop(); })("touchstart", function ModalComponent_Template_div_touchstart_1_listener() { return ctx.moveOnTop(); });
        ɵngcc0.ɵɵelementStart(3, "div", 3, 4);
        ɵngcc0.ɵɵlistener("mousedown", function ModalComponent_Template_div_mousedown_3_listener($event) { return ctx.initDrag($event); })("touchstart", function ModalComponent_Template_div_touchstart_3_listener($event) { return ctx.initDrag($event); });
        ɵngcc0.ɵɵelementStart(5, "div", 5);
        ɵngcc0.ɵɵprojection(6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "div", 6);
        ɵngcc0.ɵɵtemplate(8, ModalComponent_i_8_Template, 1, 4, "i", 7);
        ɵngcc0.ɵɵelementStart(9, "i", 8, 9);
        ɵngcc0.ɵɵlistener("click", function ModalComponent_Template_i_click_9_listener() { return ctx.hide(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "div", 10, 11);
        ɵngcc0.ɵɵprojection(13, 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "div", 12, 13);
        ɵngcc0.ɵɵprojection(16, 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵstyleProp("display", ctx.visible && ctx.backdrop ? "block" : "none");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵstyleProp("display", ctx.visible ? "block" : "none");
        ɵngcc0.ɵɵproperty("south", true)("east", true)("southEast", true)("southWest", true)("west", true)("northWest", true)("north", true)("northEast", true)("dragEventTarget", ctx.dragEventTarget)("inViewport", ctx.inViewport);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵproperty("ngIf", ctx.maximizable);
    } }, directives: [ɵngcc1.ResizableDirective, ɵngcc2.DraggableDirective, ɵngcc3.NgIf, ɵngcc3.NgClass], styles: [".ui-modal *,.ui-modal :after,.ui-modal :before{box-sizing:border-box}.ui-modal,.ui-modal-overlay{z-index:10}.ui-modal-overlay{background-color:rgba(0,0,0,.2);height:100%;width:100%}.ui-modal,.ui-modal-overlay{display:none;left:0;position:fixed;top:0}.ui-modal{background-color:#fff;box-shadow:0 .25rem .5rem 0 rgba(0,0,0,.2),0 .375rem 1.25rem 0 rgba(0,0,0,.19);min-height:12.5rem;min-width:16.25rem;outline:none;padding:0;width:31.25rem}.ui-modal-header{-webkit-user-select:none;align-items:center;background-color:var(--dt-color-primary,#5b9bd5);color:#fff;display:flex;flex-direction:row;flex-wrap:nowrap;padding:.5rem 1rem;position:relative;user-select:none}.ui-modal-body{-webkit-overflow-scrolling:touch;max-height:calc(100vh - 12.5rem);overflow-y:auto;padding:.625rem 1rem;position:relative}.ui-modal-footer{padding:1rem}.ui-titlebar{flex-grow:1;font-size:1.125rem;height:100%;overflow:hidden}.ui-controlbar,.ui-titlebar{align-items:center;display:flex}.ui-controlbar{background-color:inherit}.ui-icon{cursor:pointer;font-size:1.4rem;margin-left:.3em}.ui-icon:hover{opacity:.75}.dragging{-webkit-user-select:none;box-shadow:0 .25rem .5rem rgba(102,175,233,.6),0 .375rem 1.25rem rgba(0,0,0,.2);cursor:move;outline:0;user-select:none}.dt-icon-maximize{background-color:initial;border:.1em solid;border-top:.2em solid}.dt-icon-maximize,.dt-icon-normalize{display:inline-block;height:1em;position:relative;width:1em}.dt-icon-normalize{background-color:inherit}.dt-icon-normalize:after,.dt-icon-normalize:before{background-color:inherit;border:.1em solid;content:\"\";height:.8em;position:absolute;width:.8em}.dt-icon-normalize:before{right:0;top:0}.dt-icon-normalize:after{border-top-width:.2em;bottom:0;left:0}.dt-icon-close{background-color:initial;display:inline-block;height:1em;position:relative;width:1em}.dt-icon-close:after,.dt-icon-close:before{background-color:currentColor;content:\"\";height:.18em;left:50%;position:absolute;top:50%;width:1.2em}.dt-icon-close:before{transform:translate(-50%,-50%) rotate(-225deg)}.dt-icon-close:after{transform:translate(-50%,-50%) rotate(225deg)}.resize-handle-e{cursor:e-resize;height:100%;position:absolute;right:-.3125rem;top:0;width:.4375rem}.resize-handle-se{bottom:0;cursor:se-resize;height:1rem;position:absolute;right:0;width:1rem}.resize-handle-s{bottom:-.3125rem;cursor:s-resize;height:.4375rem;left:0;position:absolute;width:100%}.resize-handle-sw{bottom:0;cursor:sw-resize;height:7px;height:15px;left:0;position:absolute;width:15px}.resize-handle-w{cursor:w-resize;height:100%;left:-5px;position:absolute;top:0;width:7px}.resize-handle-nw{cursor:nw-resize;height:7px;height:15px;left:0;position:absolute;top:0;width:15px}.resize-handle-n{cursor:n-resize;height:7px;left:0;position:absolute;top:-5px;width:95%}.resize-handle-ne{cursor:ne-resize;height:7px;height:15px;position:absolute;right:0;top:0;width:15px}.resizing{-webkit-user-select:none;user-select:none}"], encapsulation: 2 });
ModalComponent.ctorParameters = () => [
    { type: ElementRef }
];
ModalComponent.propDecorators = {
    scrollTopEnable: [{ type: Input }],
    maximizable: [{ type: Input }],
    backdrop: [{ type: Input }],
    inViewport: [{ type: Input }],
    closeModal: [{ type: Output }],
    modalRoot: [{ type: ViewChild, args: ['modalRoot', { static: false },] }],
    modalBody: [{ type: ViewChild, args: ['modalBody', { static: false },] }],
    modalHeader: [{ type: ViewChild, args: ['modalHeader', { static: false },] }],
    modalFooter: [{ type: ViewChild, args: ['modalFooter', { static: false },] }],
    closeIcon: [{ type: ViewChild, args: ['closeIcon', { static: false },] }],
    onKeyDown: [{ type: HostListener, args: ['keydown.esc', ['$event'],] }],
    onWindowResize: [{ type: HostListener, args: ['window:resize',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ModalComponent, [{
        type: Component,
        args: [{
                selector: 'app-modal',
                template: "<div class=\"ui-modal-overlay\" [style.display]=\"(visible && backdrop) ? 'block' : 'none'\"></div>\n\n<div class=\"ui-modal\" tabindex=\"-1\" role=\"dialog\"\n     #modalRoot\n     appResizable\n     [south]=\"true\"\n     [east]=\"true\"\n     [southEast]=\"true\"\n     [southWest]=\"true\"\n     [west]=\"true\"\n     [northWest]=\"true\"\n     [north]=\"true\"\n     [northEast]=\"true\"\n     (resizing)=\"onResize($event)\"\n     appDraggable\n     [dragEventTarget]=\"dragEventTarget\"\n     [inViewport]=\"inViewport\"\n     [style.display]=\"visible ? 'block' : 'none'\"\n     (mousedown)=\"moveOnTop()\"\n     (touchstart)=\"moveOnTop()\">\n    <div class=\"ui-modal-header\" #modalHeader\n         (mousedown)=\"initDrag($event)\"\n         (touchstart)=\"initDrag($event)\">\n      <div class=\"ui-titlebar\">\n          <ng-content select=\".app-modal-header\"></ng-content>\n      </div>\n      <div class=\"ui-controlbar\">\n          <i class=\"ui-icon\"\n             *ngIf=\"maximizable\"\n             (click)=\"toggleMaximize($event)\"\n             [ngClass]=\"{'dt-icon-maximize': !maximized, 'dt-icon-normalize': maximized}\">\n          </i>\n          <i class=\"ui-icon dt-icon-close\" #closeIcon (click)=\"hide()\">\n          </i>\n      </div>\n    </div>\n\n    <div class=\"ui-modal-body\" #modalBody>\n      <ng-content select=\".app-modal-body\"></ng-content>\n    </div>\n    <div class=\"ui-modal-footer\" #modalFooter>\n      <ng-content select=\".app-modal-footer\"></ng-content>\n    </div>\n</div>\n",
                encapsulation: ViewEncapsulation.None,
                styles: [".ui-modal *,.ui-modal :after,.ui-modal :before{box-sizing:border-box}.ui-modal,.ui-modal-overlay{z-index:10}.ui-modal-overlay{background-color:rgba(0,0,0,.2);height:100%;width:100%}.ui-modal,.ui-modal-overlay{display:none;left:0;position:fixed;top:0}.ui-modal{background-color:#fff;box-shadow:0 .25rem .5rem 0 rgba(0,0,0,.2),0 .375rem 1.25rem 0 rgba(0,0,0,.19);min-height:12.5rem;min-width:16.25rem;outline:none;padding:0;width:31.25rem}.ui-modal-header{-webkit-user-select:none;align-items:center;background-color:var(--dt-color-primary,#5b9bd5);color:#fff;display:flex;flex-direction:row;flex-wrap:nowrap;padding:.5rem 1rem;position:relative;user-select:none}.ui-modal-body{-webkit-overflow-scrolling:touch;max-height:calc(100vh - 12.5rem);overflow-y:auto;padding:.625rem 1rem;position:relative}.ui-modal-footer{padding:1rem}.ui-titlebar{flex-grow:1;font-size:1.125rem;height:100%;overflow:hidden}.ui-controlbar,.ui-titlebar{align-items:center;display:flex}.ui-controlbar{background-color:inherit}.ui-icon{cursor:pointer;font-size:1.4rem;margin-left:.3em}.ui-icon:hover{opacity:.75}.dragging{-webkit-user-select:none;box-shadow:0 .25rem .5rem rgba(102,175,233,.6),0 .375rem 1.25rem rgba(0,0,0,.2);cursor:move;outline:0;user-select:none}.dt-icon-maximize{background-color:initial;border:.1em solid;border-top:.2em solid}.dt-icon-maximize,.dt-icon-normalize{display:inline-block;height:1em;position:relative;width:1em}.dt-icon-normalize{background-color:inherit}.dt-icon-normalize:after,.dt-icon-normalize:before{background-color:inherit;border:.1em solid;content:\"\";height:.8em;position:absolute;width:.8em}.dt-icon-normalize:before{right:0;top:0}.dt-icon-normalize:after{border-top-width:.2em;bottom:0;left:0}.dt-icon-close{background-color:initial;display:inline-block;height:1em;position:relative;width:1em}.dt-icon-close:after,.dt-icon-close:before{background-color:currentColor;content:\"\";height:.18em;left:50%;position:absolute;top:50%;width:1.2em}.dt-icon-close:before{transform:translate(-50%,-50%) rotate(-225deg)}.dt-icon-close:after{transform:translate(-50%,-50%) rotate(225deg)}.resize-handle-e{cursor:e-resize;height:100%;position:absolute;right:-.3125rem;top:0;width:.4375rem}.resize-handle-se{bottom:0;cursor:se-resize;height:1rem;position:absolute;right:0;width:1rem}.resize-handle-s{bottom:-.3125rem;cursor:s-resize;height:.4375rem;left:0;position:absolute;width:100%}.resize-handle-sw{bottom:0;cursor:sw-resize;height:7px;height:15px;left:0;position:absolute;width:15px}.resize-handle-w{cursor:w-resize;height:100%;left:-5px;position:absolute;top:0;width:7px}.resize-handle-nw{cursor:nw-resize;height:7px;height:15px;left:0;position:absolute;top:0;width:15px}.resize-handle-n{cursor:n-resize;height:7px;left:0;position:absolute;top:-5px;width:95%}.resize-handle-ne{cursor:ne-resize;height:7px;height:15px;position:absolute;right:0;top:0;width:15px}.resizing{-webkit-user-select:none;user-select:none}"]
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { scrollTopEnable: [{
            type: Input
        }], backdrop: [{
            type: Input
        }], closeModal: [{
            type: Output
        }], onKeyDown: [{
            type: HostListener,
            args: ['keydown.esc', ['$event']]
        }], onWindowResize: [{
            type: HostListener,
            args: ['window:resize']
        }], maximizable: [{
            type: Input
        }], inViewport: [{
            type: Input
        }], modalRoot: [{
            type: ViewChild,
            args: ['modalRoot', { static: false }]
        }], modalBody: [{
            type: ViewChild,
            args: ['modalBody', { static: false }]
        }], modalHeader: [{
            type: ViewChild,
            args: ['modalHeader', { static: false }]
        }], modalFooter: [{
            type: ViewChild,
            args: ['modalFooter', { static: false }]
        }], closeIcon: [{
            type: ViewChild,
            args: ['closeIcon', { static: false }]
        }] }); })();
