// @circlon/angular-tree-component/esm2015/lib/models/tree-dragged-element.model.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/models/tree-dragged-element.model.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as ɵngcc0 from '@angular/core';
export class TreeDraggedElement {
    constructor() {
        this._draggedElement = null;
    }
    /**
     * @param {?} draggedElement
     * @return {?}
     */
    set(draggedElement) {
        this._draggedElement = draggedElement;
    }
    /**
     * @return {?}
     */
    get() {
        return this._draggedElement;
    }
    /**
     * @return {?}
     */
    isDragging() {
        return !!this.get();
    }
}
TreeDraggedElement.ɵfac = function TreeDraggedElement_Factory(t) { return new (t || TreeDraggedElement)(); };
/** @nocollapse */ TreeDraggedElement.ɵprov = i0.ɵɵdefineInjectable({ factory: function TreeDraggedElement_Factory() { return new TreeDraggedElement(); }, token: TreeDraggedElement, providedIn: "root" });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(TreeDraggedElement, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return []; }, null); })();
if (false) {
    /** @type {?} */
    TreeDraggedElement.prototype._draggedElement;
}
