// keycloak-angular/esm2015/lib/core/interceptors/keycloak-bearer.interceptor.js
import { Injectable } from '@angular/core';
import { from } from 'rxjs';
import { mergeMap } from 'rxjs/operators';
import { KeycloakService } from '../services/keycloak.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../services/keycloak.service';
export class KeycloakBearerInterceptor {
    constructor(keycloak) {
        this.keycloak = keycloak;
    }
    isUrlExcluded({ method, url }, { urlPattern, httpMethods }) {
        const httpTest = httpMethods.length === 0 ||
            httpMethods.join().indexOf(method.toUpperCase()) > -1;
        const urlTest = urlPattern.test(url);
        return httpTest && urlTest;
    }
    intercept(req, next) {
        const { enableBearerInterceptor, excludedUrls } = this.keycloak;
        if (!enableBearerInterceptor) {
            return next.handle(req);
        }
        const shallPass = excludedUrls.findIndex(item => this.isUrlExcluded(req, item)) > -1;
        if (shallPass) {
            return next.handle(req);
        }
        return from(this.keycloak.isLoggedIn()).pipe(mergeMap((loggedIn) => loggedIn
            ? this.handleRequestWithTokenHeader(req, next)
            : next.handle(req)));
    }
    handleRequestWithTokenHeader(req, next) {
        return this.keycloak.addTokenToHeader(req.headers).pipe(mergeMap(headersWithBearer => {
            const kcReq = req.clone({ headers: headersWithBearer });
            return next.handle(kcReq);
        }));
    }
}
KeycloakBearerInterceptor.ɵfac = function KeycloakBearerInterceptor_Factory(t) { return new (t || KeycloakBearerInterceptor)(ɵngcc0.ɵɵinject(ɵngcc1.KeycloakService)); };
KeycloakBearerInterceptor.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: KeycloakBearerInterceptor, factory: KeycloakBearerInterceptor.ɵfac });
KeycloakBearerInterceptor.ctorParameters = () => [
    { type: KeycloakService }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(KeycloakBearerInterceptor, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.KeycloakService }]; }, null); })();
