// @circlon/angular-tree-component/esm2015/lib/components/tree-node-wrapper.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/tree-node-wrapper.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewEncapsulation } from '@angular/core';
import { TreeNode } from '../models/tree-node.model';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from './tree-node-expander.component';
import * as ɵngcc3 from '../directives/tree-drag.directive';
import * as ɵngcc4 from '../directives/tree-drop.directive';
import * as ɵngcc5 from './tree-node-content.component';
import * as ɵngcc6 from './tree-node-checkbox.component';

function TreeNodeWrapperComponent_div_0_tree_node_checkbox_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "tree-node-checkbox", 4);
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("node", ctx_r1.node);
} }
function TreeNodeWrapperComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    const _r3 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵtemplate(1, TreeNodeWrapperComponent_div_0_tree_node_checkbox_1_Template, 1, 1, "tree-node-checkbox", 3);
    ɵngcc0.ɵɵelement(2, "tree-node-expander", 4);
    ɵngcc0.ɵɵelementStart(3, "div", 5);
    ɵngcc0.ɵɵlistener("click", function TreeNodeWrapperComponent_div_0_Template_div_click_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r3); const ctx_r2 = ɵngcc0.ɵɵnextContext(); return ctx_r2.node.mouseAction("click", $event); })("dblclick", function TreeNodeWrapperComponent_div_0_Template_div_dblclick_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r3); const ctx_r4 = ɵngcc0.ɵɵnextContext(); return ctx_r4.node.mouseAction("dblClick", $event); })("mouseover", function TreeNodeWrapperComponent_div_0_Template_div_mouseover_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r3); const ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.node.mouseAction("mouseOver", $event); })("mouseout", function TreeNodeWrapperComponent_div_0_Template_div_mouseout_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r3); const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.node.mouseAction("mouseOut", $event); })("contextmenu", function TreeNodeWrapperComponent_div_0_Template_div_contextmenu_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r3); const ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.node.mouseAction("contextMenu", $event); })("treeDrop", function TreeNodeWrapperComponent_div_0_Template_div_treeDrop_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r3); const ctx_r8 = ɵngcc0.ɵɵnextContext(); return ctx_r8.node.onDrop($event); })("treeDropDragOver", function TreeNodeWrapperComponent_div_0_Template_div_treeDropDragOver_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r3); const ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.node.mouseAction("dragOver", $event); })("treeDropDragLeave", function TreeNodeWrapperComponent_div_0_Template_div_treeDropDragLeave_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r3); const ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.node.mouseAction("dragLeave", $event); })("treeDropDragEnter", function TreeNodeWrapperComponent_div_0_Template_div_treeDropDragEnter_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r3); const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.node.mouseAction("dragEnter", $event); });
    ɵngcc0.ɵɵelement(4, "tree-node-content", 6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵstyleProp("padding-left", ctx_r0.node.getNodePadding());
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.node.options.useCheckbox);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("node", ctx_r0.node);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵclassProp("node-content-wrapper-active", ctx_r0.node.isActive)("node-content-wrapper-focused", ctx_r0.node.isFocused);
    ɵngcc0.ɵɵproperty("treeAllowDrop", ctx_r0.node.allowDrop)("allowDragoverStyling", ctx_r0.node.allowDragoverStyling())("treeDrag", ctx_r0.node)("treeDragEnabled", ctx_r0.node.allowDrag());
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("node", ctx_r0.node)("index", ctx_r0.index)("template", ctx_r0.templates.treeNodeTemplate);
} }
const _c0 = function (a0, a1, a2, a3) { return { $implicit: a0, node: a1, index: a2, templates: a3 }; };
export class TreeNodeWrapperComponent {
    constructor() {
    }
}
TreeNodeWrapperComponent.ɵfac = function TreeNodeWrapperComponent_Factory(t) { return new (t || TreeNodeWrapperComponent)(); };
TreeNodeWrapperComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: TreeNodeWrapperComponent, selectors: [["tree-node-wrapper"]], inputs: { node: "node", index: "index", templates: "templates" }, decls: 2, vars: 8, consts: [["class", "node-wrapper", 3, "padding-left", 4, "ngIf"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], [1, "node-wrapper"], [3, "node", 4, "ngIf"], [3, "node"], [1, "node-content-wrapper", 3, "treeAllowDrop", "allowDragoverStyling", "treeDrag", "treeDragEnabled", "click", "dblclick", "mouseover", "mouseout", "contextmenu", "treeDrop", "treeDropDragOver", "treeDropDragLeave", "treeDropDragEnter"], [3, "node", "index", "template"]], template: function TreeNodeWrapperComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, TreeNodeWrapperComponent_div_0_Template, 5, 15, "div", 0);
        ɵngcc0.ɵɵelementContainer(1, 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.templates.treeNodeWrapperTemplate);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.templates.treeNodeWrapperTemplate)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction4(3, _c0, ctx.node, ctx.node, ctx.index, ctx.templates));
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgTemplateOutlet, ɵngcc2.TreeNodeExpanderComponent, ɵngcc3.TreeDragDirective, ɵngcc4.TreeDropDirective, ɵngcc5.TreeNodeContent, ɵngcc6.TreeNodeCheckboxComponent], encapsulation: 2 });
/** @nocollapse */
TreeNodeWrapperComponent.ctorParameters = () => [];
TreeNodeWrapperComponent.propDecorators = {
    node: [{ type: Input }],
    index: [{ type: Input }],
    templates: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(TreeNodeWrapperComponent, [{
        type: Component,
        args: [{
                selector: 'tree-node-wrapper',
                encapsulation: ViewEncapsulation.None,
                template: `
      <div *ngIf="!templates.treeNodeWrapperTemplate" class="node-wrapper" [style.padding-left]="node.getNodePadding()">
          <tree-node-checkbox *ngIf="node.options.useCheckbox" [node]="node"></tree-node-checkbox>
          <tree-node-expander [node]="node"></tree-node-expander>
          <div class="node-content-wrapper"
               [class.node-content-wrapper-active]="node.isActive"
               [class.node-content-wrapper-focused]="node.isFocused"
               (click)="node.mouseAction('click', $event)"
               (dblclick)="node.mouseAction('dblClick', $event)"
               (mouseover)="node.mouseAction('mouseOver', $event)"
               (mouseout)="node.mouseAction('mouseOut', $event)"
               (contextmenu)="node.mouseAction('contextMenu', $event)"
               (treeDrop)="node.onDrop($event)"
               (treeDropDragOver)="node.mouseAction('dragOver', $event)"
               (treeDropDragLeave)="node.mouseAction('dragLeave', $event)"
               (treeDropDragEnter)="node.mouseAction('dragEnter', $event)"
               [treeAllowDrop]="node.allowDrop"
               [allowDragoverStyling]="node.allowDragoverStyling()"
               [treeDrag]="node"
               [treeDragEnabled]="node.allowDrag()">

              <tree-node-content [node]="node" [index]="index" [template]="templates.treeNodeTemplate">
              </tree-node-content>
          </div>
      </div>
      <ng-container
              [ngTemplateOutlet]="templates.treeNodeWrapperTemplate"
              [ngTemplateOutletContext]="{ $implicit: node, node: node, index: index, templates: templates }">
      </ng-container>
  `
            }]
    }], function () { return []; }, { node: [{
            type: Input
        }], index: [{
            type: Input
        }], templates: [{
            type: Input
        }] }); })();
if (false) {
    /** @type {?} */
    TreeNodeWrapperComponent.prototype.node;
    /** @type {?} */
    TreeNodeWrapperComponent.prototype.index;
    /** @type {?} */
    TreeNodeWrapperComponent.prototype.templates;
}
