// @circlon/angular-tree-component/esm2015/lib/directives/tree-drag.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/tree-drag.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostListener, Input, NgZone, Renderer2 } from '@angular/core';
import { TreeDraggedElement } from '../models/tree-dragged-element.model';
/** @type {?} */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../models/tree-dragged-element.model';
const DRAG_OVER_CLASS = 'is-dragging-over';
export class TreeDragDirective {
    /**
     * @param {?} el
     * @param {?} renderer
     * @param {?} treeDraggedElement
     * @param {?} ngZone
     */
    constructor(el, renderer, treeDraggedElement, ngZone) {
        this.el = el;
        this.renderer = renderer;
        this.treeDraggedElement = treeDraggedElement;
        this.ngZone = ngZone;
        this.dragEventHandler = this.onDrag.bind(this);
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        /** @type {?} */
        let el = this.el.nativeElement;
        this.ngZone.runOutsideAngular((/**
         * @return {?}
         */
        () => {
            el.addEventListener('drag', this.dragEventHandler);
        }));
    }
    /**
     * @return {?}
     */
    ngDoCheck() {
        this.renderer.setAttribute(this.el.nativeElement, 'draggable', this.treeDragEnabled ? 'true' : 'false');
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        /** @type {?} */
        let el = this.el.nativeElement;
        el.removeEventListener('drag', this.dragEventHandler);
    }
    /**
     * @param {?} ev
     * @return {?}
     */
    onDragStart(ev) {
        // setting the data is required by firefox
        ev.dataTransfer.setData('text', ev.target.id);
        this.treeDraggedElement.set(this.draggedElement);
        if (this.draggedElement.mouseAction) {
            this.draggedElement.mouseAction('dragStart', ev);
        }
    }
    /**
     * @param {?} ev
     * @return {?}
     */
    onDrag(ev) {
        if (this.draggedElement.mouseAction) {
            this.draggedElement.mouseAction('drag', ev);
        }
    }
    /**
     * @return {?}
     */
    onDragEnd() {
        if (this.draggedElement.mouseAction) {
            this.draggedElement.mouseAction('dragEnd');
        }
        this.treeDraggedElement.set(null);
    }
}
TreeDragDirective.ɵfac = function TreeDragDirective_Factory(t) { return new (t || TreeDragDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.TreeDraggedElement), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone)); };
TreeDragDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: TreeDragDirective, selectors: [["", "treeDrag", ""]], hostBindings: function TreeDragDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("dragstart", function TreeDragDirective_dragstart_HostBindingHandler($event) { return ctx.onDragStart($event); })("dragend", function TreeDragDirective_dragend_HostBindingHandler() { return ctx.onDragEnd(); });
    } }, inputs: { draggedElement: ["treeDrag", "draggedElement"], treeDragEnabled: "treeDragEnabled" } });
/** @nocollapse */
TreeDragDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: TreeDraggedElement },
    { type: NgZone }
];
TreeDragDirective.propDecorators = {
    draggedElement: [{ type: Input, args: ['treeDrag',] }],
    treeDragEnabled: [{ type: Input }],
    onDragStart: [{ type: HostListener, args: ['dragstart', ['$event'],] }],
    onDragEnd: [{ type: HostListener, args: ['dragend',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(TreeDragDirective, [{
        type: Directive,
        args: [{
                selector: '[treeDrag]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc1.TreeDraggedElement }, { type: ɵngcc0.NgZone }]; }, { 
    /**
     * @param {?} ev
     * @return {?}
     */
    onDragStart: [{
            type: HostListener,
            args: ['dragstart', ['$event']]
        }], 
    /**
     * @return {?}
     */
    onDragEnd: [{
            type: HostListener,
            args: ['dragend']
        }], draggedElement: [{
            type: Input,
            args: ['treeDrag']
        }], treeDragEnabled: [{
            type: Input
        }] }); })();
if (false) {
    /** @type {?} */
    TreeDragDirective.prototype.draggedElement;
    /** @type {?} */
    TreeDragDirective.prototype.treeDragEnabled;
    /**
     * @type {?}
     * @private
     */
    TreeDragDirective.prototype.dragEventHandler;
    /**
     * @type {?}
     * @private
     */
    TreeDragDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    TreeDragDirective.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    TreeDragDirective.prototype.treeDraggedElement;
    /**
     * @type {?}
     * @private
     */
    TreeDragDirective.prototype.ngZone;
}
